### ISSUE TITLE
Healthy DB connections are thrown away after transaction serialization failure

### ISSUE THREAD

**Issue by eric-christian** [rails/rails#43130]
Hello :)

Last week we updated from 6.1.3.2 to 6.1.4.1 and noticed that there are many connections dropped and reestablished now.
We also got ActiveRecord::ConnectionNotEstablished within nested transactions in some special cases. (Explained in more detail at the bottom)

After some digging, i noticed that connections are thrown away after an ActiveRecord::TransactionRollbackError is raised. Which is the parent class for ActiveRecord::SerializationFailure and ActiveRecord::Deadlocked.
At least for the ActiveRecord::SerializationFailure exception, this behavior is giving us some trouble. 

We are using the transaction isolation level "serializable", where SerializationFailure's are a common thing.

We are currently using CockroachDB but as far as i know neither mysql nor postgresql leaves the connection in an undefined state after a SerializationFailure. Simply rollback and retry the transaction on the same connection is a valid way to go. So dropping the connection is not necessary. If there is a case with a DB where dropping the connection is actually required, can we maybe only drop the connection in that case?

https://github.com/rails/rails/compare/v6.1.3.2...v6.1.4#diff-353f4f101b1bb04938d761b6e29c88e0f147084e9e276854a77ae5d9a9a3b2a8R3
https://github.com/rails/rails/compare/v6.1.3.2...v6.1.4#diff-dd0ea9dfea1fd486f2b1b5eac8c798f839467173a1c2315d3ad04f4098f7dda3R333

I don't know if this was meant to be a performance improvement or actually a real issue with some databases that i don't know of. But skipping a ROLLBACK and dropping a connection are two different things. I think, but i'm not 100% sure about it right now, that the ROLLBACK [SAVEPOINT ...] is still required after a SerializationFailure.
I think you have to signal the DB that you want to ROLLBACK the failed transaction if you attempt to retry it on the same connection. This probably depends on the DB? If there is different behavior, maybe leave it to the adapter to decide what way to go? So far i don't understand why this changes were made.

**Example time! :)**
We are currently running two different implementations of retryable transactions.

1. A basic one with simple exponential backoff, which still works but issues unnecessary reconnects.
```ruby
  # simplified code
  simple_retryable_transaction do
    return yield if within_transaction? # no retries for nested transactions

    loop do
      break ActiveRecord::Base.connection.transaction { yield } # RealTransaction
    rescue ActiveRecord::SerializationFailure => error      
      #  update metrics / APM to see contention
      raise unless more_retries?
      sleep exponential_backoff_seconds
   end
 end
```

2. A more specialised and faster one, where a savepoint is rolled back. Probably only featured by CockroachDB? This one is now broken. After the first SerializationFailure the Inner- and Outer-Transaction are not on the same connection anymore.

```ruby
  # simplified code
  advanced_retryable_transaction do
    return yield if within_transaction? 

    ActiveRecord::Base.connection.transaction(joinable: false) do
      loop do
       # on the second iteration, this will create a RealTransaction on a new connection.
        break ActiveRecord::Base.connection.transaction(requires_new: true) { yield }
      rescue ActiveRecord::SerializationFailure => error      
        #  update metrics / APM to see contention
        # since the DB is prioritising the session with the most retries during contention evaluation:
        #   no sleep and no limited amount of retries
     end
     # If we had a SerializationFailure during the loop which we solved by retrying,
     #   this transaction is now attempting to issue a COMMIT on a closed connection \o/
   end
end
```

If required i'll try to put together some examples with https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#create-an-executable-test-case

Best regards,
Eric

### Expected behavior
Keep the connection after ActiveRecord::SerializationFailure and probably even after ActiveRecord::Deadlocked.

### Actual behavior
Connections are thrown away after ActiveRecord::SerializationFailure and ActiveRecord::Deadlocked.

### System configuration
**Rails version**: 6.1.4.1

**Ruby version**: 2.7.3

**Comment by ghiculescu** [comment]
> [v6.1.3.2...v6.1.4diff-353f4f101b1bb04938d761b6e29c88e0f147084e9e276854a77ae5d9a9a3b2a8R3](https://github.com/rails/rails/compare/v6.1.3.2...v6.1.4diff-353f4f101b1bb04938d761b6e29c88e0f147084e9e276854a77ae5d9a9a3b2a8R3)
> [v6.1.3.2...v6.1.4diff-dd0ea9dfea1fd486f2b1b5eac8c798f839467173a1c2315d3ad04f4098f7dda3R333](https://github.com/rails/rails/compare/v6.1.3.2...v6.1.4diff-dd0ea9dfea1fd486f2b1b5eac8c798f839467173a1c2315d3ad04f4098f7dda3R333)

These links aren't working for me, are you sure they're the right link?

**Comment by eric-christian** [comment]
Ah sorry, i simply copy&pasted them without testing them!
I wanted to point out the changes within the comparison-view-thingy but somehow github won't let me do that. 

Here are the links to the lines that i initially tried to link to:
https://github.com/rails/rails/blob/18707ab17fa492eb25ad2e8f9818a320dc20b823/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L324
https://github.com/rails/rails/blob/18707ab17fa492eb25ad2e8f9818a320dc20b823/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L335

Best regards,
Eric

**Comment by ghiculescu** [comment]
Hmmmm, does the change in https://github.com/rails/rails/pull/40541 look like the cause of your issues?

**Comment by eric-christian** [comment]
Yes, looks like it.

Is there a missunderstanding between a failed rollback and a failed transaction?
I'm fine with dropping the connection if the rollback fails or if something raises during the attempted rollback, because that would leave the session in an undefined state.
But not when a transaction fails with a serialisation failure, that is legit behavior with a defined state.
Am i missing something?

Best regards,
Eric

**Comment by ghiculescu** [comment]
@caueguerra or @jahfer, any thoughts on this one?

**Comment by eric-christian** [comment]
Any updates on this?

**Comment by jahfer** [comment]
As you already found, we throw away the connection if it hasn't been rolled back successfully:

https://github.com/rails/rails/blob/18707ab17fa492eb25ad2e8f9818a320dc20b823/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L335

Forgive my naive knowledge of `ActiveRecord::SerializationFailure`, but I would have assumed `transaction.state.rolledback?` would be `true`, given the call to `#rollback_transaction` in the transaction rescue:

https://github.com/rails/rails/blob/5d9ee652013e527129792016c103ee1133106276/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L322-L327

What we were seeing in https://github.com/rails/rails/pull/40541 was that a new exception inside of the rollback left the transaction open and left the connection in the pool. Do you know what state the transaction has when you enter the `ensure` block of `#within_new_transaction` for a `SerializationFailure`?

**Comment by eric-christian** [comment]
Hi @jahfer,

Thank you for your feedback.

From my understanding, `transaction.state.rolledback?` will only be true  if `transaction.rollback` has been called at some point.
But that is not the case because there is this line `transaction.rollback unless transaction.state.invalidated?` here 
https://github.com/rails/rails/blob/5d9ee652013e527129792016c103ee1133106276/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L308-L314

`#rollback_transaction` does not raise any exceptions within my example code. It successfully issues a `ROLLBACK` statement to the database.
And in that case, `transaction.state.rolledback?` is true within the `ensure` block. At least that was the case in the past.

 I think that in the situation where an exception is actually raised *within* a rollback, it is totally valid to drop the connection because something went bonkers. Because a rollback should always be possible as long as there is a transaction. At least that is what i know about databases and transactions. But again, `ActiveRecord::TransactionRollbackError` is not raised when a `ROLLBACK` fails, but when a transaction is required to `ROLLBACK` and tried again.

What you found within https://github.com/rails/rails/pull/40541 really seems to have been a problem with the error handling within `#within_new_transaction` or more exactly with the `ensure` block not taking into account that the `rescue` block might have been unable to successfully issue a `ROLLBACK` statement. And i think the changes to the `ensure` block are legit. While not 100%. There are still possibilities that the `ROLLBACK` at this line https://github.com/rails/rails/blob/5d9ee652013e527129792016c103ee1133106276/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L338 or this line https://github.com/rails/rails/blob/5d9ee652013e527129792016c103ee1133106276/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L353 within the `ensure` block fails without the connection being dropped.

Still, to my understanding this entire line https://github.com/rails/rails/blob/5d9ee652013e527129792016c103ee1133106276/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L324 and the `unless ....` part at this line https://github.com/rails/rails/blob/5d9ee652013e527129792016c103ee1133106276/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L311  are not necessary to fix the problem in https://github.com/rails/rails/pull/40541. Maybe the entire `transaction.state.invalidated` state is not necessary.

To my understanding `ActiveRecord::TransactionRollbackError` was never part of the problem nor should be part of the solution to https://github.com/rails/rails/pull/40541 . So i wonder why it showed up here. Just because of the name? Or was there a situation where it was encountered and led to trouble with the `ROLLBACK`? Maybe i am missing something? Maybe a special behavior of a database  that i don't know of, where a `ROLLBACK` after a serialisation error is invalid? But that doesn't seem to match with the description of the problem in https://github.com/rails/rails/pull/40541 where a `commit` was actually possible later on. 🤷‍♂️ 


Best regards,
Eric

**Comment by jahfer** [comment]
Apologies for the delay. This took a little bit of tracing down, but it looks like the issue you're describing was introduced in https://github.com/rails/rails/commit/57053ad20b40099b060fc5356831fae41764f1d5 (cc @rafaelfranca).

The line added in #40541 that throws away the connection:

https://github.com/rails/rails/blob/18707ab17fa492eb25ad2e8f9818a320dc20b823/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L335

…should actually avoid doing so if `transaction.state.rolledback?` _or_ `transaction.state.invalidated?`. This is because invalidated transactions are already rolled back at the DB level, so there's no risk of accidentally committing them (at least this is my very poor understanding of `ActiveRecord::SerializationFailure` and `ActiveRecord::Deadlocked`). 

The `invalidated` state was introduced in https://github.com/rails/rails/pull/30922 and merged after #40541.

Does my description make sense to you as well, @eric-christian?

**Comment by zarqman** [comment]
@eric-christian, I've made a pass at fixing this in #44127. Thanks for your detailed commentary above--it was of great help in confirming my own research into the issue. I'd love to hear any feedback you have on that PR and if it successfully addresses the issue for you.
